### ISSUE TITLE
关于梯度填充，fig2的复现问题

### ISSUE THREAD

**Issue by YML0823** [XuanL17/Mjolnir#3]
Hello, bothering again. I've reproduced the gradient fill operation in fig2 but I'm still getting the issue. Could you please take a look at the code, is there a reproduction of the operation mentioned in the paper?

```
import os
import numpy as np
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torchvision import datasets, transforms

# 定义超参数
batch_size = 64
learning_rate = 0.001
epochs = 5
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")


# 数据加载
def load_data(batch_size):
    transform = transforms.Compose([transforms.ToTensor(), transforms.Normalize((0.5,), (0.5,))])
    train_dataset = datasets.FashionMNIST(
        root='/home/ubuntu/Mjolnir/Mjolnir_GradientDiffusion/dataset/fashion/raw/FashionMNIST', train=True,
        download=True, transform=transform)
    test_dataset = datasets.FashionMNIST(
        root='/home/ubuntu/Mjolnir/Mjolnir_GradientDiffusion/dataset/fashion/raw/FashionMNIST', train=False,
        download=True, transform=transform)
    train_loader = torch.utils.data.DataLoader(train_dataset, batch_size=batch_size, shuffle=True)
    test_loader = torch.utils.data.DataLoader(test_dataset, batch_size=batch_size, shuffle=False)
    return train_loader, test_loader


# 定义简易的 CNN 模型
class SimpleCNN(nn.Module):
    def __init__(self):
        super(SimpleCNN, self).__init__()
        self.conv1 = nn.Conv2d(1, 32, kernel_size=3, stride=1, padding=1)
        self.conv2 = nn.Conv2d(32, 64, kernel_size=3, stride=1, padding=1)
        self.fc1 = nn.Linear(64 * 7 * 7, 128)
        self.fc2 = nn.Linear(128, 10)

    def forward(self, x):
        x = F.relu(self.conv1(x))
        x = F.max_pool2d(x, 2)
        x = F.relu(self.conv2(x))
        x = F.max_pool2d(x, 2)
        x = x.view(-1, 64 * 7 * 7)
        x = F.relu(self.fc1(x))
        x = self.fc2(x)
        return x


# 训练函数
import os


def train(model, device, train_loader, optimizer, epoch, grad_save_orig):
    model.train()
    all_grads = []

    # 确保保存路径存在
    os.makedirs(grad_save_orig[0], exist_ok=True)  # 对于清洁梯度保存路径
    os.makedirs(grad_save_orig[1], exist_ok=True)  # 对于噪声梯度保存路径

    for batch_idx, (data, target) in enumerate(train_loader):
        data, target = data.to(device), target.to(device)
        optimizer.zero_grad()
        output = model(data)
        loss = F.cross_entropy(output, target)
        loss.backward()

        if epoch == epochs and batch_idx == len(train_loader) - 1:
            # 展平每一层的梯度并将所有梯度拼接起来
            clean_gradients = {name: p.grad.clone().detach() for name, p in model.named_parameters() if
                               p.grad is not None}
            for name, grad in clean_gradients.items():
                grad_flat = grad.view(-1)  # 展平梯度
                all_grads.append(grad_flat)

            # 将所有层的梯度拼接成一个大的梯度
            all_grads = torch.cat(all_grads, dim=0)  # 拼接成一个大的向量

            # 填充梯度
            L = all_grads.size(0)  # 获取总大小
            g = int(np.ceil(np.sqrt(L)))  # 计算填充后的大小
            P = g * g - L  # 计算填充的大小

            # 填充梯度并调整为统一的形状
            all_grads_padded = F.pad(all_grads, (0, P), value=0)  # 填充零
            all_grads_padded = all_grads_padded.view(1, g, g)  # 调整为统一的形状

            # 保存填充后的清洁梯度
            np.save(f"{grad_save_orig[0]}/clean_grad_{epoch}_{batch_idx}.npy", all_grads_padded.cpu().numpy())

            # 生成噪声梯度并进行相同处理
            noisy_gradients = {name: grad + torch.randn_like(grad) * 0.1 for name, grad in clean_gradients.items()}
            noisy_grads = []
            for name, grad in noisy_gradients.items():
                grad_flat = grad.view(-1)  # 展平梯度
                noisy_grads.append(grad_flat)

            noisy_grads = torch.cat(noisy_grads, dim=0)
            noisy_grads_padded = F.pad(noisy_grads, (0, P), value=0)  # 填充零
            noisy_grads_padded = noisy_grads_padded.view(1, g, g)  # 调整为统一的形状

            # 保存填充后的噪声梯度
            np.save(f"{grad_save_orig[1]}/noisy_grad_{epoch}_{batch_idx}.npy", noisy_grads_padded.cpu().numpy())

        optimizer.step()

        if batch_idx % 100 == 0:
            print(
                f"Train Epoch: {epoch} [{batch_idx * len(data)}/{len(train_loader.dataset)} ({100. * batch_idx / len(train_loader):.0f}%)]\tLoss: {loss.item():.6f}")


# 主函数
def main():
    # 加载数据
    train_loader, test_loader = load_data(batch_size)

    # 创建模型
    model = SimpleCNN().to(device)
    optimizer = optim.Adam(model.parameters(), lr=learning_rate)

    # 梯度保存路径
    grad_save_orig = ["/home/ubuntu/Mjolnir/Mjolnir_GradientDiffusion/grads/orig",
                      "/home/ubuntu/Mjolnir/Mjolnir_GradientDiffusion/grads/noise"]
    os.makedirs(grad_save_orig[0], exist_ok=True)
    os.makedirs(grad_save_orig[1], exist_ok=True)

    # 训练和测试
    for epoch in range(1, epochs + 1):
        train(model, device, train_loader, optimizer, epoch, grad_save_orig)
        # test(model, device, test_loader)  # You can uncomment this if needed


if __name__ == "__main__":
    main()

```
